### PR TITLE
Java compiler

### PR DESCRIPTION
**Pull request by psbang**
Created a Dafny to Java compiler. Passes all test cases in Test/comp. Work in progress on passing all other test cases.

### CODE REVIEW

**Comment by samuelgruetter** [comment]
It seems you added all `DafnyRuntime` files twice: 
* in `Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/`, as well as
* in ` Test/JavaCompiler/src/DafnyClasses/`

It would be better to only have one copy of them (in `Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/`) and to use `Binaries/DafnyRuntime-1.jar` to run the tests.

**Comment by seanmcl** [comment]
Let's close this and re-open when all this is sorted out.

**Comment by samuelgruetter** [comment]
I reopened it to make sure new commits pushed to the java-compiler branch show up here and can be reviewed & discussed.

File: Source/Dafny/Compiler-java.cs [hunk 1]
```diff
@@ -475,8 +475,7 @@ protected class ClassWriter : IClassWriter {
     }
 
     private void EmitSuppression(TextWriter wr) {
-      wr.WriteLine("@SuppressWarnings(\"unchecked\")");
-      wr.WriteLine("@SuppressWarnings(\"deprecation\")");
+      wr.WriteLine("@SuppressWarnings({\"unchecked\", \"deprecation\"})");
```
Comment (samuelgruetter) [inline Source/Dafny/Compiler-java.cs]: Suppressing deprecation warnings might be an ok temporary fix, but did you find out which deprecated API you're using? The travis log:
```
+Note: ./_System/__default.java uses or overrides a deprecated API.
+Note: Recompile with -Xlint:deprecation for details.
+Note: /home/travis/build/dafny-lang/dafny/Test/comp/Extern/Library/MyInt.java uses or overrides a deprecated API.
+Note: Recompile with -Xlint:deprecation for details.
+Note: /home/travis/build/dafny-lang/dafny/Test/comp/Extern/_System/__default.java uses or overrides a deprecated API.
+Note: Recompile with -Xlint:deprecation for details.
```
suggests using `-Xlint:deprecation` to find that out, did you try that?

File: Source/DafnyDriver/DafnyDriver.cs [hunk 2]
```diff
@@ -543,10 +566,17 @@ static string WriteDafnyProgramToFiles(string dafnyProgramName, string targetPro
           otherFiles.Add(wr.Filename, sw.ToString());
         }
       }
+      string baseName = Path.GetFileNameWithoutExtension(dafnyProgramName);
       string callToMain = null;
       if (hasMain) {
         using (var wr = new TargetWriter(0)) {
+          if (DafnyOptions.O.CompileTarget is DafnyOptions.CompilationTarget.Java) {
+            wr.Write("public class {0} {{\n\t", baseName);
```
Comment (RustanLeino) [inline Source/DafnyDriver/DafnyDriver.cs]: Do a WriteLine instead of using `\n`, because depending on the system, lines are ended with `\n`, `\r`, or `\r\n`.

Also, use spaces instead of tabs.

**Comment by samuelgruetter** [comment]
Two minor but easy points to fix:

Override these members to get rid of the warnings:
```
2>C:\Users\kggruett\Documents\git\Dafny2Java\Dafny\Source\Dafny\Compiler-java.cs(32,12,32,25): warning CS0108: 'JavaCompiler.DafnySetClass' hides inherited member 'Compiler.DafnySetClass'. Use the new keyword if hiding was intended.
2>C:\Users\kggruett\Documents\git\Dafny2Java\Dafny\Source\Dafny\Compiler-java.cs(33,12,33,30): warning CS0108: 'JavaCompiler.DafnyMultiSetClass' hides inherited member 'Compiler.DafnyMultiSetClass'. Use the new keyword if hiding was intended.
2>C:\Users\kggruett\Documents\git\Dafny2Java\Dafny\Source\Dafny\Compiler-java.cs(34,12,34,25): warning CS0108: 'JavaCompiler.DafnySeqClass' hides inherited member 'Compiler.DafnySeqClass'. Use the new keyword if hiding was intended.
2>C:\Users\kggruett\Documents\git\Dafny2Java\Dafny\Source\Dafny\Compiler-java.cs(35,12,35,25): warning CS0108: 'JavaCompiler.DafnyMapClass' hides inherited member 'Compiler.DafnyMapClass'. Use the new keyword if hiding was intended.
```

`Source/DafnyRuntime/DafnyRuntimeJava/target` should be gitignored.

**Comment by samuelgruetter** [comment]
Recording two more TODO items here:

### Deprecation warnings

If I run `mvn package` in `Source/DafnyRuntime/DafnyRuntimeJava`, I get these warnings:

```
[WARNING] /C:/Users/kggruett/Documents/git/Dafny2Java/Dafny/Source/DafnyRuntime/DafnyRuntimeJava/src/test/java/BigOrdinalTest.java: Some input files use or override a deprecated API.
[WARNING] /C:/Users/kggruett/Documents/git/Dafny2Java/Dafny/Source/DafnyRuntime/DafnyRuntimeJava/src/test/java/BigOrdinalTest.java: Recompile with -Xlint:deprecation for details.
[WARNING] /C:/Users/kggruett/Documents/git/Dafny2Java/Dafny/Source/DafnyRuntime/DafnyRuntimeJava/src/test/java/TupleTest.java: Some input files use unchecked or unsafe operations.
[WARNING] /C:/Users/kggruett/Documents/git/Dafny2Java/Dafny/Source/DafnyRuntime/DafnyRuntimeJava/src/test/java/TupleTest.java: Recompile with -Xlint:unchecked for details.
```

If I run ` mvn package -Dmaven.compiler.showDeprecation=true` to get more details, I get warnings of the form

```
[WARNING] /C:/Users/kggruett/Documents/git/Dafny2Java/Dafny/Source/DafnyRuntime/DafnyRuntimeJava/src/test/java/SetTest.java:[10,30] junit.framework.Assert in junit.framework has been deprecated
```

so you should stop using this deprecated API.


### Run tests for DafnyRuntime on CI

The CI build should run `mvn test` in `Source/DafnyRuntime/DafnyRuntimeJava`.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UInt.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    public UInt not(){
+        return new UInt(~inner);
+    }
+
+    public UInt shiftLeft(int i){
+        return new UInt(inner << i);
+    }
+
+    public UInt shiftRight(int i){
+        return new UInt(inner >> i);
```
Comment (RustanLeino) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UInt.java]: This should be `>>>`. Please include the following tests (either in `Test/dafny0/Bitvectors.dfy` or in a new test file):

```
method Main() {
  var u: bv32 := 0xFFFF_FFFF;
  var v := u >> 1;
  Print([u, v]);  // 0xFFFF_FFFF 0x7FFF_FFFF
}

method Print(s: seq<bv32>) {
  var i := 0;
  while i < |s| {
    PrintAsHex(s[i] as uint);
    print " ";
    i := i + 1;
  }
  print "\n";
}

type uint = x | 0 <= x < 0x1_0000_0000

method PrintAsHex(u: uint) {
  var s, i, u := "", 0, u;
  while i < 8 {
    var d := u % 16;
    u := u / 16;
    var ch := if d < 10 then '0' + d as char else 'A' + (d - 10) as char;
    if i == 4 {
      s := "_" + s;
    }
    s := [ch] + s;
    i := i + 1;
  }
  print "0x", s;
}
```

The same comment applies to ULong.java, UShort.java, and UByte.java.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Tuple2.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        return sb.toString();
+    }
+
+    @Override
+    public int hashCode() {
+        // GetHashCode method (Uses the djb2 algorithm)
+        // https://stackoverflow.com/questions/1579721/why-are-5381-and-33-so-important-in-the-djb2-algorithm
+        long hash = 5381;
+        hash = ((hash << 5) + hash) + 0;
+        hash = ((hash << 5) + hash) + ((long) this._0.hashCode());
```
Comment (RustanLeino) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Tuple2.java]: The code in `toString()` is written to anticipate `_0` and `_1` being `null`, but the code in `hashCode()` and in `equals` will crash in such circumstances. If you expect `null`, then change `hashCode()` and `equals()`. If you don't expect `null`, change `toString()`. In either case, document this somewhere in the class.

The same comment goes for Tuple3.java.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UByte.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    public static int compare(UByte x, UByte y){
+        return Integer.compareUnsigned(x.inner, y.inner);
+    }
+
+    public int compareTo(UByte other){
+        return Integer.compareUnsigned(inner, other.inner);
+    }
+
+    public double doubleValue(){
+        return (double) inner;
```
Comment (RustanLeino) [inline Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UByte.java]: Why is this included? `double` and `float` are never used, right?

File: Source/Dafny/Compiler.cs [hunk 6]
```diff
@@ -198,19 +211,33 @@ protected interface IClassWriter {
...
     protected virtual void DeclareOutCollector(string collectorVarName, TargetWriter wr) { }  // called only for return-style calls
+    protected virtual void DeclareSpecificOutCollector(string collectorVarName, TargetWriter wr, int outCount, List<Type> types, Method m) {DeclareOutCollector(collectorVarName, wr); } // for languages that don't allow "let" or "var" expressions
     protected virtual bool UseReturnStyleOuts(Method m, int nonGhostOutCount) => false;
+    protected virtual BlockTargetWriter EmitMethodReturns(Method m, BlockTargetWriter wr) { return wr; } // for languages that need explicit return statements not provided by Dafny
     protected virtual bool SupportsMultipleReturns { get => false; }
     protected virtual bool NeedsCastFromTypeParameter { get => false; }
+    protected virtual bool SupportsAmbiguousTypeDecl { get => true; }
+    protected virtual bool FieldsInTraits { get => true; } // True if language's "trait" equivalent allows for non-final field declarations, false otherwise
+    protected bool MemberSelectObjIsTrait; // Example: if MemberSelect expression is T.x, this will be true if T is a trait, false otherwise
```
Comment (RustanLeino) [inline Source/Dafny/Compiler.cs]: Which `MemberSelect` expression is this referring to?

This would be better as some kind of parameter to `EmitMemberSelect` and `EmitAssignment`.

File: Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UByte.java (hunk 5 above)
Comment (psbang) [reply Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/UByte.java]: They were added just in case they were ever needed.
